### PR TITLE
Optimize `ImmutableMatrix`

### PR DESCRIPTION
**Pull request by siefkenj**
There was quadratic complexity growth in `ImmutableDenseMatrix`
because `list(self.args[2])` was called every time `self._mat`
was accessed.  This patch passes the request directly through.

Tests show a small speedup for small matrices and a large speedup for big matrices.

Running

```
X = DeferredVector("X")
for i in range(1,11):
    w = 100*i
    mat = Matrix(1, w, lambda i, j: X[i*j])
    print(w)
    %timeit lambdify(X, mat)
```

produces 

```
100
100 loops, best of 3: 4.73 ms per loop
200
100 loops, best of 3: 9.19 ms per loop
300
100 loops, best of 3: 13.9 ms per loop
400
100 loops, best of 3: 18.9 ms per loop
500
10 loops, best of 3: 24.2 ms per loop
600
10 loops, best of 3: 29.3 ms per loop
700
10 loops, best of 3: 34.7 ms per loop
800
10 loops, best of 3: 40.6 ms per loop
900
10 loops, best of 3: 46.4 ms per loop
1000
10 loops, best of 3: 53 ms per loop
```

on master and

```
100
100 loops, best of 3: 4.55 ms per loop
200
100 loops, best of 3: 8.34 ms per loop
300
100 loops, best of 3: 12.3 ms per loop
400
100 loops, best of 3: 16.2 ms per loop
500
10 loops, best of 3: 20 ms per loop
600
10 loops, best of 3: 23.9 ms per loop
700
10 loops, best of 3: 27.8 ms per loop
800
10 loops, best of 3: 31.7 ms per loop
900
10 loops, best of 3: 35.7 ms per loop
1000
10 loops, best of 3: 39.9 ms per loop
```

with this PR.

With a matrix of size 10000, the difference is `396 ms` vs `1.55 s`

### CODE REVIEW

**Comment by siefkenj** [comment]
This partially addresses #12731 but doesn't get it fully up to the original speed.

**Comment by isuruf** [comment]
Here's a patch that further increases the speed.
```diff
diff --git a/sympy/matrices/common.py b/sympy/matrices/common.py
index b9c68e1..21b3dcf 100644
--- a/sympy/matrices/common.py
+++ b/sympy/matrices/common.py
@@ -103,11 +103,9 @@ def entry(i, j):
                                          lambda i, j: entry(i, j))
 
     def _eval_extract(self, rowsList, colsList):
-        mat = list(self)
-        cols = self.cols
-        indices = (i * cols + j for i in rowsList for j in colsList)
-        return self._new(len(rowsList), len(colsList),
-                         list(mat[i] for i in indices))
+        def entry(i, j):
+            return self[rowsList[i], colsList[j]]
+        return self._new(len(rowsList), len(colsList), entry)
 
     def _eval_get_diag_blocks(self):
         sub_blocks = []
```

**Comment by siefkenj** [comment]
@isuruf That patch actually slows things down significantly.  I got

```
100
100 loops, best of 3: 6.51 ms per loop
200
100 loops, best of 3: 12.6 ms per loop
300
100 loops, best of 3: 18.5 ms per loop
400
10 loops, best of 3: 24.3 ms per loop
500
10 loops, best of 3: 30.5 ms per loop
600
10 loops, best of 3: 36.5 ms per loop
700
10 loops, best of 3: 43.1 ms per loop
800
10 loops, best of 3: 48.7 ms per loop
900
10 loops, best of 3: 54.6 ms per loop
1000
10 loops, best of 3: 60.2 ms per loop
```

with the above benchmark.

**Comment by siefkenj** [comment]
Thanks to @isuruf 's suggestion, I optimized `_eval_extract` and got another 15% speed increase.

```
100
100 loops, best of 3: 3.77 ms per loop
200
100 loops, best of 3: 7.04 ms per loop
300
100 loops, best of 3: 10.3 ms per loop
400
100 loops, best of 3: 13.5 ms per loop
500
100 loops, best of 3: 16.9 ms per loop
600
10 loops, best of 3: 19.9 ms per loop
700
10 loops, best of 3: 23.5 ms per loop
800
10 loops, best of 3: 26.7 ms per loop
900
10 loops, best of 3: 30.1 ms per loop
1000
10 loops, best of 3: 33.3 ms per loop
```

**Comment by siefkenj** [comment]
@jksuom Could you review this?

File: sympy/matrices/dense.py [hunk 1]
```diff
@@ -27,6 +27,17 @@ def _iszero(x):
     return x.is_zero
 
 
+def _compare_sequence(a, b):
+    """Compares the elements of a list/tupe `a`
```
Comment (jksuom) [inline sympy/matrices/dense.py]: tupe -> tuple

**Comment by jksuom** [comment]
This looks good.

**Pull request merged**